chalice/local/internal.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+"""
+This is intended only for `pytest-chalice`.
+
+https://github.com/studio3104/pytest-chalice
+"""
+
+import json
+import re
+
+from logging import getLogger
+from typing import Any, Callable, Dict, Optional  # noqa
+
+from chalice import Chalice  # noqa
+from chalice.config import Config
+from chalice.local import LocalGateway
+from chalice.local import EventType, HeaderType  # noqa
+
+logger = getLogger(__name__)
+
+# Python 3.4 or older don't have JSONDecodeError within json module
+try:
+    JSONDecodeError = json.JSONDecodeError  # type: ignore
+except AttributeError:
+    JSONDecodeError = ValueError  # type: ignore
+
+
+class InternalLocalGateway(LocalGateway):
+    def __init__(self, *args, **kwargs):
+        # type: (Any, Any) -> None
+        self.__custom_context = {}  # type: Dict
+        super(InternalLocalGateway, self).__init__(*args, **kwargs)
+
+    @property
+    def custom_context(self):
+        # type: () -> Dict[str, Any]
+        return self.__custom_context
+
+    @custom_context.setter
+    def custom_context(self, context):
+        # type: (Dict[str, Any]) -> None
+        self.__custom_context = context
+
+    def _generate_lambda_event(self, method, path, headers, body):
+        # type: (str, str, HeaderType, Optional[str])-> EventType
+        event = super(
+            InternalLocalGateway, self)._generate_lambda_event(
+                method, path, headers, body)
+        event['requestContext'].update(self.custom_context)
+        return event
+
+
+UPPERCASE_PATTERN = re.compile('([A-Z])')
+
+
+class ResponseHandler(object):
+    def __init__(self, values):
+        # type: (Dict[str, Any]) -> None
+        self.values = {}  # type: Dict
+
+        for key, value in values.items():
+            snake_key = re.sub(
+                UPPERCASE_PATTERN, lambda x: '_' + x.group(1).lower(), key)
+            self.values[snake_key] = value
+
+        try:
+            self.values['json'] = json.loads(self.values['body'])
+        except JSONDecodeError:  # type: ignore
+            logger.info(
+                'Response body is NOT JSON decodable: %s',
+                self.values['body'])
+
+    def __getattr__(self, key):
+        # type: (str) -> Any
+        try:
+            return self.values[key]
+        except KeyError:
+            raise AttributeError(
+                "'%s' object has no attribute '%s'" % (
+                    self.__class__.__name__, key))
+
+
+class RequestHandler(object):
```

Comment:
I think having an explicitly named type such as `TestHTTPClient` would be helpful instead of `RequestHandler`.

I'd would like to eventually add support for testing other chalice resources (`@app.lambda_function()`, `@app.schedule()`, `@app.on_s3_event()`), so having an explicit type of `TestHTTPClient` would be great.

Reply:
Renamed the class as you suggested!  https://github.com/aws/chalice/pull/1193/commits/2f783df65ba59ad7ff962d3780c550e0a70b6138